### PR TITLE
fix: revalidate user token with no refresh page

### PR DESCRIPTION
**Pull request by Himself65**
### Summary
🔒🔄📊

This pull request enhances the token refresh and reconnection logic for the web app and the workspace packages. It uses a custom event and a storage change slot to notify components and providers of token changes, and refactors the API functions to use a check function instead of a get function. It also improves the logging and code organization.

> _Oh, we're the web app sailors and we work with tokens fine_
> _We use the `storageChangeSlot` to keep them all in line_
> _We fetch and revalidate our data with the `useSWR` hook_
> _And we reconnect our websockets when the token changes look_

### Walkthrough
*  Move `RequestError` class from `@affine/workspace/affine/api` module to `apps/web/src/hooks/affine/use-affine-refresh-auth-token.ts` module to avoid circular dependencies ([link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bf5f61f88a4804e707ee3148aab1f66097706e910d76cca9c143776c4e69f9f3L2-R4))
*  Add `storageChangeSlot` event emitter to `@affine/workspace/affine/login` module to notify when login storage has been updated with a new token ([link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-b26da29407eb55cf368e47f2527c7466423f02cddf6767ff7afffd2033940166R83-R108))
*  Import `storageChangeSlot` event emitter in `apps/web/src/hooks/affine/use-affine-refresh-auth-token.ts`, `packages/workspace/src/affine/sync.ts`, and `packages/workspace/src/providers/index.ts` modules to listen and react to login storage changes ([link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bb7de48fbcb67e05215593f019a85c5530235ff46ba5e8c48151e05995131c5fR7), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bed295fce6582af3dda7271c0a8e3a5732ee2faf7f05ee64425e271141e42811L6-R16), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-f7ea0d405010dded0eab312a37e67b255ab2c56c8238cb311c9c2681e9688ffeL3-R6))
*  Add `checkLoginStorage` function to `@affine/workspace/affine/login` module to check and refresh token if expired before returning login storage ([link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-b26da29407eb55cf368e47f2527c7466423f02cddf6767ff7afffd2033940166R83-R108))
*  Use `checkLoginStorage` function instead of `getLoginStorage` function in all API functions in `@affine/workspace/affine/api` module to ensure valid token usage ([link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bf5f61f88a4804e707ee3148aab1f66097706e910d76cca9c143776c4e69f9f3L54-R53), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bf5f61f88a4804e707ee3148aab1f66097706e910d76cca9c143776c4e69f9f3L66-R64), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bf5f61f88a4804e707ee3148aab1f66097706e910d76cca9c143776c4e69f9f3L190-R187), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bf5f61f88a4804e707ee3148aab1f66097706e910d76cca9c143776c4e69f9f3L207-R203), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bf5f61f88a4804e707ee3148aab1f66097706e910d76cca9c143776c4e69f9f3L223-R218), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bf5f61f88a4804e707ee3148aab1f66097706e910d76cca9c143776c4e69f9f3L239-R233), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bf5f61f88a4804e707ee3148aab1f66097706e910d76cca9c143776c4e69f9f3L257-R250), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bf5f61f88a4804e707ee3148aab1f66097706e910d76cca9c143776c4e69f9f3L277-R269), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bf5f61f88a4804e707ee3148aab1f66097706e910d76cca9c143776c4e69f9f3L295-R286), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bf5f61f88a4804e707ee3148aab1f66097706e910d76cca9c143776c4e69f9f3L313-R303), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bf5f61f88a4804e707ee3148aab1f66097706e910d76cca9c143776c4e69f9f3L342-R331), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bf5f61f88a4804e707ee3148aab1f66097706e910d76cca9c143776c4e69f9f3L361-R349), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bf5f61f88a4804e707ee3148aab1f66097706e910d76cca9c143776c4e69f9f3L375-R362), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bf5f61f88a4804e707ee3148aab1f66097706e910d76cca9c143776c4e69f9f3L408-R394))
*  Replace window reload with `storageChangeSlot` event emission in `apps/web/src/hooks/affine/use-affine-refresh-auth-token.ts` module to avoid app reload when token is refreshed ([link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bb7de48fbcb67e05215593f019a85c5530235ff46ba5e8c48151e05995131c5fL24-R25))
*  Add options to `useSWR` hook in `apps/web/src/hooks/affine/use-affine-refresh-auth-token.ts` module to revalidate usage data on focus, reconnect, and mount events ([link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bb7de48fbcb67e05215593f019a85c5530235ff46ba5e8c48151e05995131c5fR39-R41))
*  Import `DebugLogger` from `@affine/debug` module in `packages/workspace/src/affine/sync.ts` module to use custom logger with namespaces and levels ([link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bed295fce6582af3dda7271c0a8e3a5732ee2faf7f05ee64425e271141e42811R1))
*  Replace `console.log` with `logger.debug` in `packages/workspace/src/affine/sync.ts` module to use `DebugLogger` with `affine-sync` namespace and `debug` level ([link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bed295fce6582af3dda7271c0a8e3a5732ee2faf7f05ee64425e271141e42811L31-R36))
*  Add `dispose` and `apis` variables to `createAffineGlobalChannel` function in `packages/workspace/src/affine/sync.ts` module to hold listener and websocket client objects ([link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bed295fce6582af3dda7271c0a8e3a5732ee2faf7f05ee64425e271141e42811L56-R62))
*  Create and dispose listener for `storageChangeSlot` event in `connect` and `disconnect` methods of websocket client object in `packages/workspace/src/affine/sync.ts` module to reconnect websocket client with new token when login storage changes ([link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-bed295fce6582af3dda7271c0a8e3a5732ee2faf7f05ee64425e271141e42811L65-R83))
*  Add `dispose` and `apis` variables to `createAffineWebSocketProvider` function in `packages/workspace/src/providers/index.ts` module to hold listener and websocket provider objects ([link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-f7ea0d405010dded0eab312a37e67b255ab2c56c8238cb311c9c2681e9688ffeL23-R30))
*  Create and dispose listener for `storageChangeSlot` event in `connect` and `disconnect` methods of websocket provider object in `packages/workspace/src/providers/index.ts` module to reconnect websocket provider with new token when login storage changes ([link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-f7ea0d405010dded0eab312a37e67b255ab2c56c8238cb311c9c2681e9688ffeR37-R40), [link](https://github.com/toeverything/AFFiNE/pull/1842/files?diff=unified&w=0#diff-f7ea0d405010dded0eab312a37e67b255ab2c56c8238cb311c9c2681e9688ffeL56-R71))

### CODE REVIEW

File: apps/web/src/hooks/affine/use-affine-refresh-auth-token.ts [hunk 1]
```diff
@@ -38,5 +36,8 @@ export function useAffineRefreshAuthToken(
   useSWR('autoRefreshToken', {
     fetcher: revalidate,
     refreshInterval,
+    revalidateOnFocus: true,
```
Comment (pengx17) [inline apps/web/src/hooks/affine/use-affine-refresh-auth-token.ts:39]: Is there a minimum revalidate throttle for these flags?

File: packages/workspace/src/affine/login.ts [hunk 2]
```diff
@@ -78,6 +80,32 @@ export const getLoginStorage = (): LoginResponse | null => {
   return null;
 };
 
+export const storageChangeSlot = new Slot();
+
+export const checkLoginStorage = async (
+  prefixUrl = '/'
+): Promise<LoginResponse> => {
+  const storage = getLoginStorage();
+  assertExists(storage, 'Login token is not set');
```
Comment (pengx17) [inline packages/workspace/src/affine/login.ts:89]: I think it should be OK to have no login token here and just return undefined Login info

File: packages/workspace/src/providers/index.ts [hunk 3]
```diff
@@ -20,13 +26,18 @@ const createAffineWebSocketProvider = (
...
-  return {
+  let dispose: Disposable | undefined = undefined;
+  const apis: AffineWebSocketProvider = {
     flavour: 'affine-websocket',
     background: false,
     cleanup: () => {
       assertExists(webSocketProvider);
       webSocketProvider.destroy();
       webSocketProvider = null;
+      dispose = storageChangeSlot.on(() => {
```
Comment (pengx17) [inline packages/workspace/src/providers/index.ts:37]: It seems this listener binding should be added outside of the cleanup 🤔

File: apps/web/src/hooks/affine/use-affine-refresh-auth-token.ts (hunk 1 above)
Comment (himself65) [reply apps/web/src/hooks/affine/use-affine-refresh-auth-token.ts:39]: https://swr.vercel.app/docs/api#options

File: packages/workspace/src/affine/login.ts (hunk 2 above)
Comment (himself65) [reply packages/workspace/src/affine/login.ts:89]: i think it's not ok since the caller do need auth otherwise it will 401 anyway

**Pull request merged**